plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsLocalIndexExporter.java

```diff
@@ -0,0 +1,216 @@
+
+    public QueryInsightsLocalIndexExporter(
+        ClusterService clusterService,
+        Client client,
+        String localIndexName,
+        InputStream localIndexMapping
+    ) {
+        super(QueryInsightsExporterType.LOCAL_INDEX, localIndexName);
+        this.clusterService = clusterService;
+        this.client = client;
+        this.localIndexMapping = localIndexMapping;
+    }
+
+    /**
+     * Export the data to the predefined local OpenSearch Index
+     *
+     * @param records the data to export
+     * @throws IOException if an error occurs
+     */
+    @Override
+    public synchronized void export(List<T> records) throws IOException {
+        if (records.size() == 0) {
+            return;
+        }
+        if (checkIfIndexExists()) {
+            bulkRecord(records);
+        } else {
+            // local index not exist
+            initLocalIndex(new ActionListener<>() {
+                @Override
+                public void onResponse(CreateIndexResponse response) {
+                    if (response.isAcknowledged()) {
+                        log.debug(
+                            String.format(Locale.ROOT, "successfully initialized local index %s for query insight.", getIdentifier())
+                        );
+                        try {
+                            bulkRecord(records);
+                        } catch (IOException e) {
+                            log.error(String.format(Locale.ROOT, "fail to ingest query insight data to local index, error: %s", e));
+                        }
+                    } else {
+                        log.error(
+                            String.format(
+                                Locale.ROOT,
+                                "request to created local index %s for query insight not acknowledged.",
+                                getIdentifier()
+                            )
+                        );
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    log.error(String.format(Locale.ROOT, "error creating local index for query insight: %s", e));
+                }
+            });
+        }
+    }
+
+    /**
+     * Util function to check if a local OpenSearch Index exists
+     *
+     * @return boolean
+     */
+    private boolean checkIfIndexExists() {
+        ClusterState clusterState = clusterService.state();
+        return clusterState.getRoutingTable().hasIndex(this.getIdentifier());
+    }
+
+    /**
+     * Initialize the local OpenSearch Index for the exporter
+     *
+     * @param listener the listener to be notified upon completion
+     * @throws IOException if an error occurs
+     */
+    private synchronized void initLocalIndex(ActionListener<CreateIndexResponse> listener) throws IOException {
+        CreateIndexRequest createIndexRequest = new CreateIndexRequest(this.getIdentifier()).mapping(getIndexMappings())
```

Comment:
The synchronized keyword should be on the method checkAndCreateIndex. The call to checkIfIndexExists should be pretty lightweight and can be easily made.

synchronized void checkAndCreate() {
  if (!checkIfIndexExists()) {
    initLocalIndex()
  }
  
}

Reply:
In the export function, there will be 2 senarios, 
1. When the index doesn't exist, create the index and when the creation finishes, bulk the record with a listener
2. When the index exist, just do bulk.

Thus we need to check if index exists even before any operations (like init index). That's the reason I'm separating those 2 functions to check if exists and init index. Do you think it make sense?

Reply:
The problem is the usage of synchronized keyword. You need to instead synchronize the whole check and create block. Hence, adding synchronized keyword on initLocalIndex method is meaningless.

Reply:
Ah that make sense! let me update the logic here.